runtime/caml/mlvalues.h

```diff
@@ -84,12 +84,34 @@ typedef opcode_t * code_t;
 #define Unsigned_long_val(x) ((uintnat)(x) >> 1)
 #define Unsigned_int_val(x)  ((int) Unsigned_long_val(x))
 
-/* Encoded exceptional return values, when functions are suffixed with
-   _exn. Encoded exceptions are invalid values and must not be seen
-   by the garbage collector. */
-#define Make_exception_result(v) ((v) | 2)
-#define Is_exception_result(v) (((v) & 3) == 2)
-#define Extract_exception(v) ((v) & ~3)
+/* A 'result' type for OCaml computations. */
+
+/* The [caml_result] type represents the result of computing an OCaml
+   term -- either a value or an exception:
+   - if [is_exception] is false,
+     the computation resulted in a value stored in [data]
+   - if [is_exception] is true,
+     the computation resulted in an exception stored in [data]
+
+   This plays a similar role to the [('a, exn) result] type in OCaml,
+   with a different representation. Returning this type, instead of
+   raising exceptions directly, lets the caller implement proper
+   cleanup and propagate the exception themselves.
+
+   To extract the value or raise the exception, use
+     value caml_get_value_or_raise(caml_result)
```

Comment:
You are indeed hinting at a non-internal definition of `caml_get_value_or_raise` (which I am in favour of)